grafana/grafana-image-renderer
# Support new concurrency mode: `contextPerRenderKey`

PR adds support for a new concurrency mode, `contextPerRenderKey` which improves the performance of the dashboard crawl process without negatively affecting the existing use cases.


With `contextPerRenderKey` mode enabled, image-renderer will reuse the same [browser context](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#class-browsercontext) (created via CDP - https://chromedevtools.github.io/devtools-protocol/tot/Target/#method-createBrowserContext) for all requests with the same `domain` and `renderKey` within a certain timeframe. 

The dashboard crawler reuses the same `renderKey` for all renders within a single crawl thanks to the [concept of RenderingSession](https://github.com/grafana/grafana/pull/44098/files#diff-f412c5dc4b56f8c250a37c17e8291f74bd0ef8694afb1f8f4ce7d77368b414a1R67). All other use cases create a unique `renderKey` for each request, causing `contextPerRenderKey` mode to behave exactly the same as the existing `context` mode

The main benefits are:
- lower setup overhead for each render: new `browserContext` vs new page in existing `browserContext`
   - impact: ~~negligible
- lower CPU/mem overhead: `browserContext` is a separate browser window
   - impact: ~~negligible
- shared JS cache
   - impact: big! but depends on the network conditions

Shared JS cache is by far the biggest deal. Tests done in local env with 94 dashboards and 6 crawler threads - results are the average of 3 runs (low variance)

| network                     | context | contextPerRenderKey |
|-----------------------------|---------|---------------------|
| not throttled               | 60s     | 60s                 |
| throttled: max 4mbps, 100ms | 95s     | 70s                 |
| throttled: max 1mbps, 100ms | 190s    | 78s                 |



### to test:

1. check out latest `main` grafana branch
2. enable `dashboardPreviews` feature flag
3. configure grafana to use http version of image-renderer via `custom.ini` config
```
[rendering]
server_url = http://localhost:8081/render
```
4. go to _server admin -> stats and license_ and click `start`
   - ![image](https://user-images.githubusercontent.com/23451458/152234755-fa7cec3c-1ea5-4eee-88a9-0e0dba01717e.png)
